File: generated-code/go/codecraft/FileReadWrite/model/debug_interface/debug_data.go
package debug_interface

import "fmt"
import "io"
import . "trans_gen_test/stream"

// Debug data can be drawn in the app
type DebugData interface {
    // Write DebugData to writer
    Write(writer io.Writer)

    // Get string representation of DebugData
    String() string
}

// Read DebugData from reader
func ReadDebugData(reader io.Reader) DebugData {
    switch ReadInt32(reader) {
    case 0:
        return ReadDebugDataLog(reader)
    case 1:
        return ReadDebugDataPrimitives(reader)
    case 2:
        return ReadDebugDataPlacedText(reader)
    }
    panic("Unexpected tag value")
}

// Log some text
type DebugDataLog struct {
    // Text to show
    Text string
}

func NewDebugDataLog(text string) DebugDataLog {
    return DebugDataLog {
        Text: text,
    }
}

// Read Log from reader
func ReadDebugDataLog(reader io.Reader) DebugDataLog {
    var text string
    text = ReadString(reader)
    return DebugDataLog {
        Text: text,
    }
}

// Write Log to writer
func (debugDataLog DebugDataLog) Write(writer io.Writer) {
    WriteInt32(writer, 0)
    text := debugDataLog.Text
    WriteString(writer, text)
}

// Get string representation of Log
func (debugDataLog DebugDataLog) String() string {
    stringResult := "{ "
    stringResult += "Text: "
    text := debugDataLog.Text
    stringResult += "\"" + text + "\""
    stringResult += " }"
    return stringResult
}

// Draw primitives
type DebugDataPrimitives struct {
    // Vertices
    Vertices []ColoredVertex
    // Primitive type
    PrimitiveType PrimitiveType
}

func NewDebugDataPrimitives(vertices []ColoredVertex, primitiveType PrimitiveType) DebugDataPrimitives {
    return DebugDataPrimitives {
        Vertices: vertices,
        PrimitiveType: primitiveType,
    }
}

// Read Primitives from reader
func ReadDebugDataPrimitives(reader io.Reader) DebugDataPrimitives {
    var vertices []ColoredVertex
    vertices = make([]ColoredVertex, ReadInt32(reader))
    for verticesIndex := range vertices {
        var verticesElement ColoredVertex
        verticesElement = ReadColoredVertex(reader)
        vertices[verticesIndex] = verticesElement
    }
    var primitiveType PrimitiveType
    primitiveType = ReadPrimitiveType(reader)
    return DebugDataPrimitives {
        Vertices: vertices,
        PrimitiveType: primitiveType,
    }
}

// Write Primitives to writer
func (debugDataPrimitives DebugDataPrimitives) Write(writer io.Writer) {
    WriteInt32(writer, 1)
    vertices := debugDataPrimitives.Vertices
    WriteInt32(writer, int32(len(vertices)))
    for _, verticesElement := range vertices {
        verticesElement.Write(writer)
    }
    primitiveType := debugDataPrimitives.PrimitiveType
    WriteInt32(writer, int32(primitiveType))
}

// Get string representation of Primitives
func (debugDataPrimitives DebugDataPrimitives) String() string {
    stringResult := "{ "
    stringResult += "Vertices: "
    vertices := debugDataPrimitives.Vertices
    stringResult += "[ "
    for verticesIndex, verticesElement := range vertices {
        if verticesIndex != 0 {
            stringResult += ", "
        }
        stringResult += verticesElement.String()
    }
    stringResult += " ]"
    stringResult += ", "
    stringResult += "PrimitiveType: "
    primitiveType := debugDataPrimitives.PrimitiveType
    stringResult += PrimitiveTypeToString(primitiveType)
    stringResult += " }"
    return stringResult
}

// Draw text
type DebugDataPlacedText struct {
    // Vertex to determine text position and color
    Vertex ColoredVertex
    // Text
    Text string
    // Text alignment (0 means left, 0.5 means center, 1 means right)
    Alignment float32
    // Font size in pixels
    Size float32
}

func NewDebugDataPlacedText(vertex ColoredVertex, text string, alignment float32, size float32) DebugDataPlacedText {
    return DebugDataPlacedText {
        Vertex: vertex,
        Text: text,
        Alignment: alignment,
        Size: size,
    }
}

// Read PlacedText from reader
func ReadDebugDataPlacedText(reader io.Reader) DebugDataPlacedText {
    var vertex ColoredVertex
    vertex = ReadColoredVertex(reader)
    var text string
    text = ReadString(reader)
    var alignment float32
    alignment = ReadFloat32(reader)
    var size float32
    size = ReadFloat32(reader)
    return DebugDataPlacedText {
        Vertex: vertex,
        Text: text,
        Alignment: alignment,
        Size: size,
    }
}

// Write PlacedText to writer
func (debugDataPlacedText DebugDataPlacedText) Write(writer io.Writer) {
    WriteInt32(writer, 2)
    vertex := debugDataPlacedText.Vertex
    vertex.Write(writer)
    text := debugDataPlacedText.Text
    WriteString(writer, text)
    alignment := debugDataPlacedText.Alignment
    WriteFloat32(writer, alignment)
    size := debugDataPlacedText.Size
    WriteFloat32(writer, size)
}

// Get string representation of PlacedText
func (debugDataPlacedText DebugDataPlacedText) String() string {
    stringResult := "{ "
    stringResult += "Vertex: "
    vertex := debugDataPlacedText.Vertex
    stringResult += vertex.String()
    stringResult += ", "
    stringResult += "Text: "
    text := debugDataPlacedText.Text
    stringResult += "\"" + text + "\""
    stringResult += ", "
    stringResult += "Alignment: "
    alignment := debugDataPlacedText.Alignment
    stringResult += fmt.Sprint(alignment)
    stringResult += ", "
    stringResult += "Size: "
    size := debugDataPlacedText.Size
    stringResult += fmt.Sprint(size)
    stringResult += " }"
    return stringResult
}

File: generated-code/python/codecraft/FileReadWrite/model/debug_interface/debug_data.py
from model.debug_interface.colored_vertex import ColoredVertex
from model.debug_interface.primitive_type import PrimitiveType
from stream_wrapper import StreamWrapper
from typing import List

class DebugData:
    """Debug data can be drawn in the app"""

    @staticmethod
    def read_from(stream: StreamWrapper) -> "DebugData":
        """Read DebugData from input stream
        """
        tag = stream.read_int()
        if tag == Log.TAG:
            return DebugData.Log.read_from(stream)
        if tag == Primitives.TAG:
            return DebugData.Primitives.read_from(stream)
        if tag == PlacedText.TAG:
            return DebugData.PlacedText.read_from(stream)
        raise Exception("Unexpected tag value")

class Log(DebugData):
    """Log some text"""

    TAG = 0

    __slots__ = ("text",)

    text: str

    def __init__(self, text: str):
        self.text = text
        """Text to show"""

    @staticmethod
    def read_from(stream: StreamWrapper) -> "Log":
        """Read Log from input stream
        """
        text = stream.read_string()
        return Log(text)
    
    def write_to(self, stream: StreamWrapper):
        """Write Log to output stream
        """
        stream.write_int(self.TAG)
        stream.write_string(self.text)
    
    def __repr__(self):
        return "Log(" + \
            repr(self.text) + \
            ")"

DebugData.Log = Log

class Primitives(DebugData):
    """Draw primitives"""

    TAG = 1

    __slots__ = ("vertices","primitive_type",)

    vertices: List[ColoredVertex]
    primitive_type: PrimitiveType

    def __init__(self, vertices: List[ColoredVertex], primitive_type: PrimitiveType):
        self.vertices = vertices
        """Vertices"""
        self.primitive_type = primitive_type
        """Primitive type"""

    @staticmethod
    def read_from(stream: StreamWrapper) -> "Primitives":
        """Read Primitives from input stream
        """
        vertices = []
        for _ in range(stream.read_int()):
            vertices_element = ColoredVertex.read_from(stream)
            vertices.append(vertices_element)
        primitive_type = PrimitiveType(stream.read_int())
        return Primitives(vertices, primitive_type)
    
    def write_to(self, stream: StreamWrapper):
        """Write Primitives to output stream
        """
        stream.write_int(self.TAG)
        stream.write_int(len(self.vertices))
        for element in self.vertices:
            element.write_to(stream)
        stream.write_int(self.primitive_type)
    
    def __repr__(self):
        return "Primitives(" + \
            repr(self.vertices) + \
            ", " + \
            repr(self.primitive_type) + \
            ")"

DebugData.Primitives = Primitives

class PlacedText(DebugData):
    """Draw text"""

    TAG = 2

    __slots__ = ("vertex","text","alignment","size",)

    vertex: ColoredVertex
    text: str
    alignment: float
    size: float

    def __init__(self, vertex: ColoredVertex, text: str, alignment: float, size: float):
        self.vertex = vertex
        """Vertex to determine text position and color"""
        self.text = text
        """Text"""
        self.alignment = alignment
        """Text alignment (0 means left, 0.5 means center, 1 means right)"""
        self.size = size
        """Font size in pixels"""

    @staticmethod
    def read_from(stream: StreamWrapper) -> "PlacedText":
        """Read PlacedText from input stream
        """
        vertex = ColoredVertex.read_from(stream)
        text = stream.read_string()
        alignment = stream.read_float()
        size = stream.read_float()
        return PlacedText(vertex, text, alignment, size)
    
    def write_to(self, stream: StreamWrapper):
        """Write PlacedText to output stream
        """
        stream.write_int(self.TAG)
        self.vertex.write_to(stream)
        stream.write_string(self.text)
        stream.write_float(self.alignment)
        stream.write_float(self.size)
    
    def __repr__(self):
        return "PlacedText(" + \
            repr(self.vertex) + \
            ", " + \
            repr(self.text) + \
            ", " + \
            repr(self.alignment) + \
            ", " + \
            repr(self.size) + \
            ")"

DebugData.PlacedText = PlacedText